src/leaderboard/leaderboard.service.ts

```diff
@@ -22,47 +20,66 @@ export class LeaderboardService {
     return this.leaderboardRepository.save(newLeaderboard);
   }
 
-  async sortLeaderboard(
-    leaderboardId: string,
-    leaderboardName: string,
-    metrics: string[],
-    sortingOrder: SortingOrders[],
-    index: number,
-    list: PlayerLeaderboard[] | undefined,
-  ): Promise<SortedResult> {
-    if (list === undefined) {
-      const list = await this.playerLeaderboardRepository.find({
-        where: {
-          leaderboardId: leaderboardId,
-        },
-      });
-    }
+  async sortLeaderboard(leaderboardId: string): Promise<any> {
+    let temp = 0;
+    let start = 0;
+    let end = 0;
+    let duplicatesIndex = 0;
+    let preSorted = [];
+    let postSorted = [];
 
-    if (sortingOrder[index].includes(SortingOrders.ASC)) {
-      list.sort((x, y) => (x.score[metrics[index]] > y.score[metrics[index]] ? -1 : 1));
-    }
-    if (sortingOrder[index].includes(SortingOrders.DESC)) {
-      list.sort((x, y) => (x.score[metrics[index]] > y.score[metrics[index]] ? 1 : -1));
-    }
+    const leaderboard: Leaderboard = await this.leaderboardRepository.findOne(leaderboardId);
+    let list = await this.playerLeaderboardRepository.find({
+      where: {
+        leaderboardId: leaderboardId,
+      },
+    });
 
-    const tempScores = [];
+    for (let i = 0; i < leaderboard.metrics.length; i++) {
+      let outcome = 1;
+      const metric = leaderboard.metrics[i];
+      const sortingOrder = leaderboard.sortingOrders[i];
+      const reverse = sortingOrder === SortingOrders.DESC ? -1 : 1;
 
-    for (const player of list) {
-      tempScores.push(player.score[metrics[index]]);
-    }
+      list.sort((a, b) => {
+        if (a.score[metric] < b.score[metric]) {
+          return reverse * -1;
+        } else if (a.score[metric] > b.score[metric]) {
+          return reverse * 1;
+        } else {
+          temp = a.score[metric];
+          duplicatesIndex++;
+          outcome = 0;
+          return outcome;
+        }
+      });
 
-    if (this.hasDuplicates(tempScores)) {
-      return this.sortLeaderboard(leaderboardId, leaderboardName, metrics, sortingOrder, index + 1, list);
-    } else {
-      const result: SortedResult = {
-        name: leaderboardName,
-        entries: list,
-      };
-      return result;
-    }
-  }
+      if (preSorted.length !== 0 && postSorted.length === 0) {
+        list.push(...preSorted.concat(list));
+        preSorted = [];
+      } else if (postSorted.length !== 0 && preSorted.length === 0) {
+        list.push(...list.concat(postSorted));
+        postSorted = [];
+      } else if (postSorted.length !== 0 && preSorted.length !== 0) {
+        list = preSorted.concat(list).concat(postSorted);
+        postSorted = [];
+        preSorted = [];
+      }
```

Comment:
those statements are entered after the first loop is done and it turned out that we have duplicated values for the first metric. In lines 73-81 I'm narrowing down the sorting range to the duplicated values, so that within the next iteration only those entries will be sorted by the next metric. Those entries (fx. players with score 2000 for executionTime) are always placed next to each other in the sorted array, which creates a range automatically. So what we need is to sort this range of players. The originally sorted array in line 44 is trimmed to contain only this range. I'm storing the cut parts in preSorted and postSorted. After the trimmed version is sorted, I'm connecting the cut parts again. Then setting them to [], so that we are ready for the next iteration with a fresh start.